### PR TITLE
Add support for AWS IRSA

### PR DESCRIPTION
**Pull request by thesuperzapper**
### What changes were proposed in this pull request?

The goal of this PR is to improve support for using S3 as the Cloud Object Store. Specifically it enables support for [IAM Roles for Service Accounts (IRSA)](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) authentication, in addition to the existing hard-coded `AWS_ACCESS_KEY_ID` (`cos_username`) and `AWS_SECRET_ACCESS_KEY` (`cos_password`) configs.

__Other Changes:__
- Make `cos_username` and `cos_password` NOT required fields for new runtimes
    - (IRSA requires that no access id/key is set) 
- Updates `minio` python client library from `6.X.X` to `7.X.X`
   - (Minio client `7.0.0` was the first version to support IRSA)
- removes `test_fail_bad_endpoint_main_method()` test
   - (Minio `7.X.X` no longer can throws an `InvalidEndpointError` error for invalid `cos-endpoint`)

### How was this pull request tested?

- Running `make test`
- In an AWS EKS cluster with IRSA enabled

[Edit] For testing set environment variables as follows prior to launching JupyterLab:
 - `ELYRA_GITHUB_ORG` to `thesuperzapper`
 - `ELYRA_GITHUB_BRANCH` to `cos-support-irsa`

 
----

Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by ptitzler** [comment]
The suggested change
- requires updates to the content in user_guide/runtime-conf.md
- should introduce an 'authentication type' selector for COS in the runtime schema (similar to what we already have for KFP authentication) 

I've already associated this PR with https://github.com/elyra-ai/elyra/issues/2338. Once delivered, 2338 should enable us to resolve the usability issue (only show/enforce properties that are applicable)

**Comment by thesuperzapper** [comment]
> The suggested change
> 
> * requires updates to the content in user_guide/runtime-conf.md

I am happy to do this change

> * should introduce an 'authentication type' selector for COS in the runtime schema (similar to what we already have for KFP authentication)

I am not sure we need to wait for this larger change to enable this feature, no existing feature should break with this change, it just allows you to NOT specify `cos_username` and `cos_password`.

Secondarily, it's probably a good idea to get this merged first, so we can update the `minio` python client library to `7.X.X` without also introducing lots of other things (keeps PRs small).

---

@ptitzler Do you think we can get this merged if I update the docs you asked?

File: elyra/airflow/bootstrapper.py
Comment (lresende) [inline elyra/airflow/bootstrapper.py:79]: Are we still passing this? I believe we removed it from input_params for security issues (leaking these details) and only pass them as ENV VARS?

File: elyra/airflow/bootstrapper.py
Comment (lresende) [inline elyra/airflow/bootstrapper.py:86]: How is the role being propagated to the container? Are you assuming that the container would have the info and use the same for all users?

File: elyra/airflow/bootstrapper.py
Comment (thesuperzapper) [reply elyra/airflow/bootstrapper.py:79]: This is for the unit tests, it's not used in production.

Here is the [`_get_operation_instance()`](https://github.com/elyra-ai/elyra/blob/67779c3917a4587bd8a9f2e589addda0fc27a64d/elyra/tests/kfp/test_bootstrapper.py#L155-L156) function which sets `cos-user` and `cos-password` used in unit tests.

File: elyra/airflow/bootstrapper.py
Comment (thesuperzapper) [reply elyra/airflow/bootstrapper.py:86]: This is how IRSA works, it injects `AWS_ROLE_ARN` and  `AWS_WEB_IDENTITY_TOKEN_FILE` based of the Kubernetes Pod's ServiceAccount.

If the user has chosen to not pass `cos_user` and `cos_password` we check if they have set up IRSA (out of band, with respect to Elyra).

Does that answer your question?

**Comment by ptitzler** [comment]
> I am not sure we need to wait for this larger change to enable this feature, no existing feature should break with this change, it just allows you to NOT specify `cos_username` and `cos_password`.
> 
> Secondarily, it's probably a good idea to get this merged first, so we can update the `minio` python client library to `7.X.X` without also introducing lots of other things (keeps PRs small).
> 
> @ptitzler Do you think we can get this merged if I update the docs you asked?

Yes certainly, but a follow-up PR (to be included in the same Elyra release) would be required to update the schema to allow for the selection of an "auth type"  (similar to what we added for [KFP authentication](https://github.com/elyra-ai/elyra/blob/master/elyra/metadata/schemas/kfp.json#L77-L86)).

The follow-up would 
 - query a new COS auth runtime config property to determine which auth type to use
 - use the new runtime config property value in the code instead of `cos_username` and `cos_secret` to determine how to manipulate the environment variables
 - add migration code for the existing runtime configuration schema ([similar to this](https://github.com/elyra-ai/elyra/blob/master/elyra/pipeline/kfp/kfp_metadata.py) that sets the COS auth runtime config property
 
If you don't think you'll have the bandwidth to make those follow-up changes let me know and I'll take care of it.

File: elyra/pipeline/kfp/processor_kfp.py
Comment (ptitzler) [inline elyra/pipeline/kfp/processor_kfp.py:513]: Did you mean to treat this as an optional property? The KFP and Airflow schema updates you've made still list `cos_endpoint` as required and the existing code `cos_endpoint = runtime_configuration.metadata['cos_endpoint']` would therefore be more appropriate.

Comment (thesuperzapper) [reply elyra/pipeline/kfp/processor_kfp.py:513]: I was making this consistent with the [`processor_airflow.py`](https://github.com/elyra-ai/elyra/blob/243dab7a22761ef71c1f34cce0bf4828cdd8e130/elyra/pipeline/airflow/processor_airflow.py#L94).

NOTE: this does not change the fact that `cos_endpoint` is still required, as it is specified under [`metadata/schemas/kfp.json`](https://github.com/elyra-ai/elyra/blob/25ffbe35965426b89bd91bbf69660959bcec6846/elyra/metadata/schemas/kfp.json#L162)

**Comment by thesuperzapper** [comment]
@ptitzler I have updated the PR with docs inside `docs/source/user_guide/runtime-conf.md`.

I believe this should be ready to merge now, unless there are other comments. 

---

Regarding any follow-up PR to add an auth mechanism like we have for Kubeflow, I would greatly appreciate if you could take this up.

**Comment by thesuperzapper** [comment]
Actually, I have found a bug, please don't merge yet.

File: elyra/pipeline/kfp/processor_kfp.py
Comment (ptitzler) [reply elyra/pipeline/kfp/processor_kfp.py:513]: Got it! I am all for consistency. However, I confirmed with Alan that the referenced Airflow code should have used `['..']` instead of `.get('')` to access required properties. Since the metadata service enforces the required constraint the existing code should never fail. Can you please revert the change to avoid the impression that this particular property is optional?

**Comment by ptitzler** [comment]
doc updates look good

**Comment by thesuperzapper** [comment]
@ptitzler I am still investigating a possible issue, please give me a couple of hours.

File: elyra/pipeline/kfp/processor_kfp.py
Comment (thesuperzapper) [reply elyra/pipeline/kfp/processor_kfp.py]: Done in https://github.com/elyra-ai/elyra/pull/2335/commits/4f057142900ccfd0c4502ac493b5549ef230dd7d

**Comment by thesuperzapper** [comment]
@ptitzler I just tested this in a real-world EKS cluster (with IRSA enabled), and after the fix in https://github.com/elyra-ai/elyra/pull/2335/commits/9347b616b20b585b3b44dcd197a9d1a098a936eb, this seems to be working properly now.

From my end, this looks good to merge.

**Comment by thesuperzapper** [comment]
@ptitzler is there anything else we need to do for this to be merged?

I have tested it in a real-world EKS cluster, and it works with IRSA if `cos_username` and `cos_password` are not set.

**Review by ptitzler — COMMENTED** [review]
LGTM - verified that kfp/aa generic pipeline execution yields the expected results for the legacy use cases:
 - uid + password (no Kubernetes secret defined)
 - uid + password (with Kubernetes secret defined)

No access to AWS - hence didn't test that part.

**Review by ptitzler — APPROVED** [review]
LGTM

**Comment by thesuperzapper** [comment]
@ptitzler I assume you also tested it with `ELYRA_GITHUB_ORG` and `ELYRA_GITHUB_BRANCH` env-vars set? Because this PR updates `elyra/airflow/bootstrapper.py`, `elyra/kfp/bootstrapper.py`, `etc/generic/requirements-elyra.txt`, and Elyra will download these from the main `elyra-ai/elyra` repo by default.
- `export ELYRA_GITHUB_ORG=thesuperzapper`
- `export ELYRA_GITHUB_BRANCH=cos-support-irsa`

___PS:__ We should probably ensure our unit tests have these variables set 
(assuming we are actually doing end-to-end Kubeflow pipelines testes in the CI/CD)_

**Comment by thesuperzapper** [comment]
@ptitzler I am happy for this to be merged if you are.

**Pull request merged**
